rclpy/src/rclpy/timer.cpp

```diff
@@ -0,0 +1,255 @@
+  }
+  int64_t elapsed_time;
+  rcl_ret_t ret = rcl_timer_get_time_since_last_call(timer, &elapsed_time);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get time since last timer call");
+  }
+
+  return elapsed_time;
+}
+
+int64_t
+get_timer_period(py::capsule pytimer)
+{
+  auto timer = static_cast<rcl_timer_t *>(
+    rclpy_handle_get_pointer_from_capsule(pytimer.ptr(), "rcl_timer_t"));
+  if (!timer) {
+    throw py::error_already_set();
+  }
+  int64_t timer_period;
+  rcl_ret_t ret = rcl_timer_get_period(timer, &timer_period);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to get timer period");
+  }
+  return timer_period;
+}
+
+void
+cancel_timer(py::capsule pytimer)
+{
+  auto timer = static_cast<rcl_timer_t *>(
+    rclpy_handle_get_pointer_from_capsule(pytimer.ptr(), "rcl_timer_t"));
+  if (!timer) {
+    throw py::error_already_set();
+  }
+  rcl_ret_t ret = rcl_timer_cancel(timer);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to cancel timer");
+  }
+}
+
+bool
+is_timer_canceled(py::capsule pytimer)
+{
+  auto timer = static_cast<rcl_timer_t *>(
+    rclpy_handle_get_pointer_from_capsule(pytimer.ptr(), "rcl_timer_t"));
+  if (!timer) {
+    throw py::error_already_set();
+  }
+  bool is_canceled;
+  rcl_ret_t ret = rcl_timer_is_canceled(timer, &is_canceled);
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to check if timer is canceled");
+  }
+  return is_canceled;
+}
+
+/// Handle destructor for timer
+static void
+_rclpy_destroy_timer(void * p)
+{
+  auto tmr = static_cast<rcl_timer_t *>(p);
+  if (!tmr) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_rclpy_destroy_timer got NULL pointer");
+    return;
+  }
+
+  rcl_ret_t ret = rcl_timer_fini(tmr);
+  if (RCL_RET_OK != ret) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini timer: %s",
+      rcl_get_error_string().str);
+  }
+  PyMem_Free(tmr);
+}
+
+py::capsule
+create_timer(py::capsule pyclock, py::capsule pycontext, int64_t period_nsec)
+{
+  auto clock = static_cast<rcl_clock_t *>(
+    rclpy_handle_get_pointer_from_capsule(pyclock.ptr(), "rcl_clock_t"));
+  if (!clock) {
+    throw py::error_already_set();
+  }
+
+  auto context = static_cast<rcl_context_t *>(
+    rclpy_handle_get_pointer_from_capsule(pycontext.ptr(), "rcl_context_t"));
+  if (!context) {
+    throw py::error_already_set();
+  }
+
+  auto deleter = [](rcl_timer_t * ptr) {_rclpy_destroy_timer(ptr);};
+  auto timer = std::unique_ptr<rcl_timer_t, decltype(deleter)>(
+    static_cast<rcl_timer_t *>(PyMem_Malloc(sizeof(rcl_timer_t))),
+    deleter);
+  if (!timer) {
+    throw std::bad_alloc();
+  }
+
+  *timer = rcl_get_zero_initialized_timer();
+
+  rcl_allocator_t allocator = rcl_get_default_allocator();
```

Comment:
Should we create an `rcl_allocator_t` to use `PyMem_Malloc` (et al.)?

Reply:
I don't know. Profiling it sounds like a good idea, but I don't know what affect it would have on performance.